vault/plugin_catalog.go

```diff
@@ -45,24 +45,36 @@ type PluginCatalog struct {
 	directory       string
 	logger          log.Logger
 
-	// externalPlugins holds plugin process connections by plugin name
+	// externalPlugins holds plugin process connections by a key which is
+	// generated from the plugin runner config.
 	//
 	// This allows plugins that suppport multiplexing to use a single grpc
 	// connection to communicate with multiple "backends". Each backend
 	// configuration using the same plugin will be routed to the existing
 	// plugin process.
-	externalPlugins map[string]*externalPlugin
+	externalPlugins map[externalPluginsKey]*externalPlugin
 	mlockPlugins    bool
 
 	lock sync.RWMutex
 }
 
+// Only plugins running with identical PluginRunner config can be multiplexed,
+// so we use the PluginRunner input as the key for the external plugins map.
+//
+// However, to be a map key, it must be comparable:
+// https://go.dev/ref/spec#Comparison_operators.
+// In particular, the PluginRunner struct has slices and a function which are not
+// comparable, so we need to marshal it into a string which is.
+type externalPluginsKey string
+
+func makeExternalPluginsKey(pluginRunner *pluginutil.PluginRunner) (externalPluginsKey, error) {
+	b, err := json.Marshal(pluginRunner)
```

Comment:
The only thing I worry about when I see this is that `encoding/json.Marshal` does not promise to preserve struct field ordering. (I can't imagine a case where it *wouldn't*, though.) So this is probably safe, but it always gives me a little hesitation.

`encoding/gob` does promise to serialize in field order, but then the key would not be even remotely human-readable.

Reply:
Hmm, good question. The values only live in-memory, so we don't need ordering to be consistent across versions of Go, we would only need ordering to be deterministic for the lifetime of one Vault process. Looking at the implementation, it certainly does look deterministic currently, but I'll dig into this one a bit more.

Reply:
Some languages have done things like purposefully randomizing key ordering when serializing or printing to "punish" those of us who rely on deterministic order :) I worry about the Go authors deciding to do that. (Though they seem pretty pragmatic about not breaking production code.)

Reply:
I slightly take it back: Go purposefully randomizes iteration order in maps https://github.com/golang/go/blob/master/src/runtime/map.go#L844

Reply:
Currently that's just for iterating over maps, though it's not beyond the realms of possibility they extend that pattern to serialization in the future. For serializing maps into JSON currently though, it sorts their keys: https://pkg.go.dev/encoding/json#Marshal

Reply:
To be on the safe side I could make the key a struct, and use json.Marshal to turn env/args into a string (as they're variable length), and use a fixed length array for the sha256. WDYT? Should be 100% safe right, as the ordering of arrays in JSON values should be guaranteed to be preserved?

Reply:
In the end I decided to use `hex.EncodeToString(p.Sha256)` for the SHA instead of a fixed length array as it's just a bit more forgiving for unforeseen situations like if we manage to not have a SHA in the catalog for some reason.

Reply:
Ordering of arrays in JSON values definitely guaranteed. I think this looks good :+1: